include/pybind11/embed.h

```diff
@@ -83,29 +85,103 @@ struct embedded_module {
     }
 };
 
+struct wide_char_arg_deleter {
+    void operator()(wchar_t* ptr) const {
+#if PY_VERSION_HEX >= 0x030500f0
+        // API docs: https://docs.python.org/3/c-api/sys.html#c.Py_DecodeLocale
+        PyMem_RawFree(ptr);
+#else
+        delete[] ptr;
+#endif
+    }
+};
+
+inline wchar_t* widen_chars(char* safe_arg) {
+#if PY_VERSION_HEX >= 0x030500f0
+    wchar_t* widened_arg = Py_DecodeLocale(safe_arg, nullptr);
+#else
+    wchar_t* widened_arg = nullptr;
+#  if defined(HAVE_BROKEN_MBSTOWCS) && HAVE_BROKEN_MBSTOWCS
+    size_t count = strlen(safe_arg);
+#  else
+    size_t count = mbstowcs(nullptr, safe_arg, 0);
+#  endif
+    if (count != static_cast<size_t>(-1)) {
+        widened_arg = new wchar_t[count + 1];
+        mbstowcs(widened_arg, safe_arg, count + 1);
+    }
+#endif
+    return widened_arg;
+}
+
+/// Python 2.x/3.x-compatible version of `PySys_SetArgv`
+inline void set_interpreter_argv(int argc, char** argv, bool add_program_dir_to_path) {
+    // Before it was special-cased in python 3.8, passing an empty or null argv
+    // caused a segfault, so we have to reimplement the special case ourselves.
+    char** safe_argv = argv;
+    std::unique_ptr<char*[]> argv_guard;
+    std::unique_ptr<char[]> argv_inner_guard;
+    if (nullptr == argv || argc <= 0) {
```

Comment:
```suggestion
    if (argv == nullptr || argc <= 0) {
```

I prefer the variable first (such as in the second half of this line, in fact).